### PR TITLE
[RFC] Lens fix metadata scaling and fuji spline

### PR DESCRIPTION
**Pull request by sgotti**
@jenshannoschwalm @FreddieWitherden
This is quite complex (at least for me) and could be totally wrong (I could have made only wrong assumptions). I'll try to explain it with some examples.
These can be split in two different PR/issues but to be able to explain this I had to keep them together.

---
**First issue**: Playing with the scaling slider introduced in #12880 I noticed something strange. Changing the scale value also changes to distortion effect.

Correction with auto calculated scale value
current_default_autocalc_scale

Correction with scale value set to 0.9
current_scale_0 9

As you can see the distortion correction is totally different.

I assumed that scaling is the same as in lensfun (scale the image) and autoscaling is just the optimal value to not see the black parts caused by distortion correction.
If that assumption is correct than the issue is that the scale value is applied during spline generation by multipling the cor_rgb spline output values. So for the same input value we'll get different output values that will change the distortion effect.

So I made the first patch **lens: fix metadata scaling logic** where, instead of affecting the spline with the scale, the scale is done in process_md (and all the related functions: modify_roi_in, distortion_transform, distortion_backtransform and
distor_mask).

Correction with auto calculated scale value
scale_fix_default_autocalc_scale

Correction with scale value set to 0.9
scale_fix_scale_0 9

OOC jpeg
ooc_jpeg

Now the scale value scales the image keeping the same distortion. 

---
So the Fuji related patch **lens: fix fuji spline calculation**:

If you look at the borders of the last fuji corrected image (after the scale patch) you'll see that it's a bit more distorted than the one without the patch (because the spline is not affected by the scale value) and also completely wrong at the corners.
Additionally you'll see the last part is linear and not curved, this is related to the images metadata of this image (I'll explain it better if you want) and will be transitively fixed by this patch.

This patch is based on a theory that could be totally wrong.

These are the fuji metadata distortion values of the example image:
```
Geometric Distortion Params     : 416.5555556 0.3535211268 0.5 0.6126760563 0.7070422535 0.7908450704 0.8661971831 0.9352112676 1 1.06056338 -0.4689941406 -1.19178772 -2.002914429 -2.89112854 -3.877334595 -4.99041748 -6.241638184 -7.568359375 -7.568359375
```

We have for radius 1 a value of -7.568359375. This was reverse engineered by @FreddieWitherden as the percentage of correction that the value requires.

**The main point is that** the metadata lens correction logic uses the distortion spline to map from a destination radius to the source radius.

My theory is that the fuji metadata values (and so the generated spline) are related to the source image radius and not the destination radius like currently done. So the formulas will be:

$R{in} = R_{out} \cdot (1 + \frac{k}{100})$

So for a radius of 1 we should "move" it to:
$R_{out} =  \frac{1}{(1 + \frac{-7.568359375}{100})} = 1.081880613$

So this patch converts the spline knots (x) from source radius to destination radius and calculates the related coefficients (y) (the algorithm is just experimental and could be changed/improved).

Correction with auto calculated scale value
scale_fuji_fix_default_autocalc_scale

Correction with scale value set to 0.9
scale_fuji_fix_scale_0 9

Additionally, I just converted the example image using adobe dng converter that adds a warprectilinear opcode 3 and it's handled by darktable thanks to PR #12880 . Well, the output image are practically identical (ignore vignetting not corrected by dng).

Correction with embedded metadata using dng warprectilinear:
dng_warprectilinear

Now this images look much better corrected and:
* It's very similar to the lensfun correction.
**But**
* **The strange thing** is that it looks also much better than the ooc jpeg (that's one of the reason why I'm not sure what I did is right since this means that the fuji ooc correction are not right/perfect).
* So, perhaps, using the ooc jpeg as a reference for reverse engineering the metadata logic isn't the right thing to do.

Additionally this patch adds a knot with no corrections if a there isn't a knot for the center (radius 0) (XTransIV/V metadata doesn't provide a 0 knot so the interpolation will distort the central part of the image more than what's needed. Injecting a 0 knot with no correction improves the center distortion.)

## Other included patches

`lens: fix _get_autoscale_md logic`:  This fixes a not perfect autoscale calculation that will lead to not using all the available image for fuji/sony corrections. We should only consider the borders when calculating the scale. See the commit description for more details.

`lens: use _get_autoscale_md value also for dng corrections`: with dng correction you can see in some images some pixels of the black border. This is probably due to some rounding errors when converting from the polynomial to the spline. So execute _get_autoscale_md also for dng correction.

@jenshannoschwalm @FreddieWitherden I'd like to know your thoughts on this.

If it looks good, after some more improvements etc.., there'll be also the need to keep backward compatibility with current edits.

### CODE REVIEW

**Comment by jenshannoschwalm** [comment]
First comments:

Indeed the scale parameter was not about scale but about "scale&distortiion" so your idea makes sense for me.

But the scale parameter for lensfun was also used this way iirc. So we would loose control of finetuning the distorion here?

For dng file i added the "scale" to be available in GUI as i was not completely sure it was correctly interpreted in the algo.

**Comment by sgotti** [comment]
> First comments:
> 
> Indeed the scale parameter was not about scale but about "scale&distortiion" so your idea makes sense for me.

Sorry I'm not sure I understand this comment. 

> 
> But the scale parameter for lensfun was also used this way iirc. 

From what I can see lensfun scale just scales the images but doesn't affect distortion (the scaling is done by lensfun methods and not by lens iop). What I tried to do with the first patch (`lens: fix metadata scaling logic`) is just do the same for metadata based correction: scale only the image.

Here it was my first doubt: is this correct (i.e. just scale like done for lensfun) or on metadata based correction scale has another meaning that I missed? In such case the scale fine tune is completely useless and wrong since it modifies the distortion correction.

The scaling with this patch is done scaling the destination radius, lowering `scale fine-tune` values it will scale the destination radius to values greater than 1 so we will map more image data. This means that we're resolving the spline for x values greater than 1.

While it works for dng (not sure about sony) (since the spline is computed by resolving the polynomial and could be improved by mapping knots to values > 1) for fuji I noticed some issues (the distortion curve that ends before the corners in the example image) and bas distortion correction than ooc jpeg.

But this can be caused by both strange metadata (since we have values also for a knot in 1.06 with the same value as knot 1), bad assumption of the metadata values (experiment in patch 2) or that just all of the above done in patch 1 is wrong...


> So we would loose control of finetuning the distorion here?
You can always fine tune (in metadata based correction, lenfun never had this option) with the `distortion fine tune` slider.

**Comment by sgotti** [comment]
@jenshannoschwalm @FreddieWitherden To add some more info I just converted the example image using adobe dng converter that adds a warprectilinear opcode 3 and it's handled by darktable thanks to PR #12880 . Well, the output image from the second patch **lens: fix fuji spline calculation** are practically identical (ignore vignetting not corrected by dng).

Correction with embedded metadata using dng warprectilinear:
dng_warprectilinear
 
Correction with embedded metadata with this PR patches:
scale_fuji_fix_default_autocalc_scale

They are both better than the ooc jpeg.

So the conversion logic of the dng converter to generate the warprectilinear polynomial data is quite similar to patch **lens: fix fuji spline calculation**. Now it could just be a broken clock that gives the right result two times a day, but perhaps there's something right in my reasoning.

Additionally, perhaps, using the ooc jpeg as a reference for reverse engineering the metadata logic isn't the right thing to do.

**Comment by FreddieWitherden** [comment]
When I was reverse engineering the encoding my goal was to match the out of camera JPEG's.  It is typically true that one can do better in terms of corrections.  A good chunk of this is due to copy variation; your copy of a lens could be quite different to mine (Sony used to be notorious here).  Thus, coming up with your own corrections is likely to yield a better result.

Given we typically do not have access to 10s of copies of a lens, trying to match the out of camera corrections is a reasonable compromise.  One would hope these corrections work well 'on average' across all copies, even if one can do better for particular lenses.  For Fuji there may also be an aspect of not wanting to totally correct due to excessive cropping of loss of resolution.

Returning to the question at hand: what is the 'true' model?  Here I think we need a lot more samples from a bunch of lenses at a range of focus distances on several different bodies (I believe they play some role in deciding what coefficients get written rather than just passing them verbatim from the lens firmware).

**Comment by sgotti** [comment]
> Returning to the question at hand: what is the 'true' model? Here I think we need a lot more samples from a bunch of lenses at a range of focus distances on several different bodies (I believe they play some role in deciding what coefficients get written rather than just passing them verbatim from the lens firmware).

For fuji I tried with different lenses and examples images around and with the updated algo in patch2 better matches the lensfun correction and (**what I think is an important point**) is quite identical to the conversion done by adobe dng converter (I think they had access to the metadata specification and algorithms by fuji). Instead the ooc jpeg is always less corrected.
I can also take different pictures of a grid to better understand distorsion (fuji xf35 f2, sigma 18-50 f2.8, fuji xc 15-45).

I have also some not yet pushed changes to better improve the fuji distortion: 
* XTransIV/V metadata doesn't provide a 0 knot so the interpolation will distort the central part of the image more than what's needed. Injecting a 0 knot with no correction improves the center distortion.
* Improve the autoscale algorithm since now it scales too much leaving out some useful image data.

---
if you have time, I'll be very happy if you can test this PR (patch2 is really raw but should work for distortion, not sure if vignetting and ca are right) and if you can answer the various questions in the first post:
* scale parameter meaning (only scale the image or also change the distortion effect and why?). From my tests I can't understand the original logic where scale affects the spline correction values. Patch1 just makes scale to scale image like done by lensfun.
* The possibility that fuji correction data are related to the source radius and so they should be converted to the destination radius (patch 2). (dng warprectilinear function just maps from the destination radius to source radius).

**Comment by sgotti** [comment]
Updated the PR with some additional commits:

`lens: fix _get_autoscale_md logic`:  This fixes a not perfect autoscale calculation that will lead to not using all the available image for fuji/sony corrections. We should only consider the borders when calculating the scale. See the commit description for more details.

`lens: use _get_autoscale_md value also for dng corrections`: with dng correction you can see in some images some pixels of the black border. This is probably due to some rounding errors when converting from the polynomial to the spline. So execute _get_autoscale_md also for dng correction.

@jenshannoschwalm If you want to give it a try I'd be quite happy. I'm shooting different fuji images to better asses corrections. Currently I'm quite happy with the latest fuji new reverse engineered logic since it gives images identical to the one done by the dng correction you implemented when converting a fuji raf to dng using the adobe dng converter.

Anyway also without the fuji related patch I think all the other changes are worth a try since I think the implemented scale logic (just scale the image) has much more sense and gives a real meaning to the scale fine tune. It's just that these without the fuji patch shows that fuji correction are worse than before and also totally wrong at the borders (that's the reason why I think the current fuji logic isn't right).

**Comment by jenshannoschwalm** [comment]
Will try this in depth this weekend. Had been too busy with other dt stuff 😀

**Comment by jenshannoschwalm** [comment]
I just spend some hours checking my dng testfiles, everything is working just right :-) Although i mostly have dng files from leica and seem to know what they are putting as opcode parameters into the exif data. (Don't see a difference between your pr and current master)

One thing i am not so sure about is - i saw a comment from you in the source - the centre may not be right in the middle. I don't know - have not seen - any files that don't do so but the dng specs define that. So we might want to take care, no idea how other vendors are taking care of...

Something else i was always not completely sure of, we do a linear spline approximation. Could we do better here? For the dng's i take all knots, tested also to use more but that didn't improve anything. Yet, i am still surprised how fuji/sony do this.

**Comment by FreddieWitherden** [comment]
>     * scale parameter meaning (only scale the image or also change the distortion effect and why?). From my tests I can't understand the original logic where scale affects the spline correction values. Patch1 just makes scale to scale image like done by lensfun.

I believe that your changes are correct and that the original algorithm I proposed/implemented is in error here.

>     * The possibility that fuji correction data are related to the source radius and so they should be converted to the destination radius (`lens: fix fuji spline calculation` patch). (dng warprectilinear function just maps from the destination radius to source radius).

One way we can interrogate this is by changing the coefficients (using exiftool) of an image and then feeding it through Adobe's DNG converter to see what happens.  If there is no change, then we can surmise that it is using a corrections database like lensfun, if there is a change its extent can help us understand what is going on (say by making a change which gives a very different correction if the source radius is used verses the destination).

**Comment by FreddieWitherden** [comment]
> Something else i was always not completely sure of, we do a linear spline approximation. Could we do better here? For the dng's i take all knots, tested also to use more but that didn't improve anything. Yet, i am still surprised how fuji/sony do this.

A better approach would be a monotonic cubic spline.  They are a bit more fiddly to code (and may not give any improvements in terms of corrections).  Details are scant online, but see p62 of the attached.

[main.pdf](https://github.com/darktable-org/darktable/files/11362013/main.pdf)

**Comment by sgotti** [comment]
@jenshannoschwalm @FreddieWitherden Thanks a lot for your responses.
I'm quite sure the Adobe DNG Cconverter uses the fuji metadata (I cannot change them with exiftool since they are not directly writable and will require manual hex editing) but the same len on different fuji camera version provided different metadata knots and also the polynomial is different. So I started some python notebooks to implement the lens algos, plot the fuji metadata and the ADC generated polynomial and compare them trying to understand how the polynomial provides a better correction. I think this is related on how the polynomial is internally auto scaled. For this reason I'm also not sure that the patch that changes the logic of the fuji spline is right, it's just similar to the ADC polynomial in some parts but they don't totally match. So I'd like to reverse engineer/understand how ADC scale the polynomial.

So, for the moment I'll put this PR on hold until I better understand these things.

The current algorithm isn't probably in error and matches the fuji ooc jpeg well. I think the confusion starts by the fact that the current scale param definition is confusing since it's not lensfun image scale but coefficient scale and perhaps both kind of scaling should be implemented but with two different parameters.

In the meantime I'll create some small PRs to fix some things I found in the current implementation.

**Comment by sgotti** [comment]
@jenshannoschwalm @FreddieWitherden After a lot of tests using a jupyter notebook I'm quite sure that the fuji metadata are related to the input radius and so they should be converted to output radius. To be really sure I also tested a popular windows software provided for free for fuji that also uses embedded metadata and the results are quite identical.

The differences from the ADC polynomial looks related to the fact that the polynomial is calculated doing a polynomial fit of the input metadata points (no cubic spline interpolation, and I can create something similar using numpy/scipy `curvefit` method). For the same reason I don't think doing an improved spline interpolation will change a lot the output result.

I'll update this PR later by doing some more changes:
* I'd like to fix the spline calculation to also consider the required image scaling to mimic dng metadata that is already scaled to fit the image. So with scale = 1 the image will be already correctly "cropped" to not see black parts. User can use the slider to scale in/out like done with lensfun. 
* If needed, avoid breaking previous edits keeping both algorithm versions and saving which version to use in the params.

**Comment by sgotti** [comment]
Opened a new PR #14468 with all the changes discovered during this analysis and to keep this as a reference.
